experimenter/primitives/dsbox_encoder.py

```diff
@@ -0,0 +1,319 @@
+
+from d3m import container
+from d3m.base import utils
+from d3m.metadata import hyperparams, params
+import d3m.metadata.base as mbase
+from d3m.metadata.base import DataMetadata
+from d3m.metadata.hyperparams import UniformInt
+from d3m.primitive_interfaces.base import CallResult
+from d3m.primitive_interfaces.unsupervised_learning import UnsupervisedLearnerPrimitiveBase
+
+from experimenter import constants
+
+_logger = logging.getLogger(__name__)
+
+Input = container.DataFrame
+Output = container.DataFrame
+
+
+class EncParams(params.Params):
+    mapping: typing.Dict
+    cat_columns: typing.List[str]
+    empty_columns: typing.List[int]
+
+
+class EncHyperparameter(hyperparams.Hyperparams):
+    n_limit = UniformInt(lower=0, upper=100, default=12,
+                         description='Limits the maximum number of columns generated from a single categorical column',
+                         semantic_types=['http://schema.org/Integer',
+                                         'https://metadata.datadrivendiscovery.org/types/TuningParameter'])
+    use_columns = hyperparams.Set(
+        elements=hyperparams.Hyperparameter[int](-1),
+        default=(),
+        semantic_types=['https://metadata.datadrivendiscovery.org/types/ControlParameter'],
+        description="A set of column indices to force primitive to operate on. If any specified column cannot be parsed, it is skipped.",
+    )
+    exclude_columns = hyperparams.Set(
+        elements=hyperparams.Hyperparameter[int](-1),
+        default=(),
+        semantic_types=['https://metadata.datadrivendiscovery.org/types/ControlParameter'],
+        description="A set of column indices to not operate on. Applicable only if \"use_columns\" is not provided.",
+    )
+    return_result = hyperparams.Enumeration(
+        values=['append', 'replace', 'new'],
+        default='replace',
+        semantic_types=['https://metadata.datadrivendiscovery.org/types/ControlParameter'],
+        description="Should parsed columns be appended, should they replace original columns, or should only parsed columns be returned? This hyperparam is ignored if use_semantic_types is set to false.",
+    )
+    use_semantic_types = hyperparams.UniformBool(
+        default=False,
+        semantic_types=['https://metadata.datadrivendiscovery.org/types/ControlParameter'],
+        description="Controls whether semantic_types metadata will be used for filtering columns in input dataframe. Setting this to false makes the code ignore return_result and will produce only the output dataframe"
+    )
+    add_index_columns = hyperparams.UniformBool(
+        default=True,
+        semantic_types=['https://metadata.datadrivendiscovery.org/types/ControlParameter'],
+        description="Also include primary index columns if input data has them. Applicable only if \"return_result\" is set to \"new\".",
+    )
+
+
+class Encoder(UnsupervisedLearnerPrimitiveBase[Input, Output, EncParams, EncHyperparameter]):
```

Comment:
Aside from the primitive metadata, I only had to change 1 file in this primitive to get it to correctly encode columns of categorical integers.